ospfd/ospf_gr.c

```diff
@@ -0,0 +1,824 @@
+	time_t timestamp = 0;
+
+	filepath = ospf_gr_nvm_filepath(ospf);
+	inst_name = ospf->name ? ospf->name : VRF_DEFAULT_NAME;
+
+	json = json_object_from_file(filepath);
+	if (json == NULL)
+		json = json_object_new_object();
+
+	json_object_object_get_ex(json, "instances", &json_instances);
+	if (!json_instances) {
+		json_instances = json_object_new_object();
+		json_object_object_add(json, "instances", json_instances);
+	}
+
+	json_object_object_get_ex(json_instances, inst_name, &json_instance);
+	if (!json_instance) {
+		json_instance = json_object_new_object();
+		json_object_object_add(json_instances, inst_name,
+				       json_instance);
+	}
+
+	json_object_object_get_ex(json_instance, "timestamp", &json_timestamp);
+	if (json_timestamp) {
+		time_t now;
+		unsigned long remaining_time;
+
+		/* Check if the grace period has already expired. */
+		now = time(NULL);
+		timestamp = json_object_get_int(json_timestamp);
+		if (now > timestamp) {
+			ospf_gr_restart_exit(
+				ospf, "grace period has expired already");
+		} else {
+			/* Schedule grace period timeout. */
+			ospf->gr_info.restart_in_progress = true;
+			remaining_time = timestamp - time(NULL);
+			if (IS_DEBUG_OSPF_GR)
+				zlog_debug(
+					"GR: remaining time until grace period expires: %lu(s)",
+					remaining_time);
+			thread_add_timer(master, ospf_gr_grace_period_expired,
+					 ospf, remaining_time,
+					 &ospf->gr_info.t_grace_period);
+		}
+	}
+
+	json_object_object_del(json_instances, inst_name);
+
+	json_object_to_file_ext(filepath, json, JSON_C_TO_STRING_PRETTY);
+	json_object_free(json);
+}
+
+/* Prepare to start a Graceful Restart. */
+static void ospf_gr_prepare(void)
+{
+	struct ospf *ospf;
+	struct ospf_interface *oi;
+	struct listnode *onode;
+
+	for (ALL_LIST_ELEMENTS_RO(om->ospf, onode, ospf)) {
+		struct listnode *inode;
+
+		if (!ospf->gr_info.restart_support
+		    || ospf->gr_info.prepare_in_progress)
+			continue;
+
+		if (IS_DEBUG_OSPF_GR)
+			zlog_debug(
+				"GR: preparing to perform a graceful restart [period %u second(s)]",
+				ospf->gr_info.grace_period);
+
+		/* Freeze OSPF routes in the RIB. */
+		if (ospf_zebra_gr_enable(ospf, ospf->gr_info.grace_period)) {
+			zlog_warn("%s: failed to activate graceful restart",
+				  __func__);
+			continue;
+		}
+
+		/* Send a Grace-LSA to all neighbors. */
+		for (ALL_LIST_ELEMENTS_RO(ospf->oiflist, inode, oi))
+			(void)ospf_gr_lsa_originate(oi);
+
+		/* Record end of the grace period in non-volatile memory. */
+		ospf_gr_nvm_update(ospf);
+
+		/*
+		 * Mark that a Graceful Restart preparation is in progress, to
+		 * prevent ospfd from flushing its self-originated LSAs on exit.
+		 */
+		ospf->gr_info.prepare_in_progress = true;
+	}
+}
+
+DEFUN(graceful_restart_prepare, graceful_restart_prepare_cmd,
+      "graceful-restart prepare ospf",
+      "Graceful Restart commands\n"
+      "Prepare upcoming graceful restart\n"
+      "Prepare to restart the OSPF process")
+{
+	ospf_gr_prepare();
```

Comment:
I think , we have to allow only when the O bit is set . i.e opaque capability is enabled , right ? Since neighbours  will process GRACE LSAs only with opaque capability negotiated neighbours .

Reply:
Good point. `ospf_gr_prepare()` now checks if `OSPF_OPAQUE_CAPABLE` is set before doing anything (and logs an error for all restarting instances that don't have the opaque capability enabled).